File: src/c04_professional_classes/CapabilityInterfaceDemo.java
package c04_professional_classes;

public class CapabilityInterfaceDemo {
    interface Flyable {
        void fly();
    }

    interface Swimmable {
        void swim();
    }

    static class Bird implements Flyable {
        @Override
        public void fly() {
            System.out.println("Bird is flying...");
        }
    }

    static class Fish implements Swimmable {
        @Override
        public void swim() {
            System.out.println("Fish is swimming...");
        }
    }

    public static void main(String[] args) {
        Bird bird = new Bird();
        Fish fish = new Fish();

        bird.fly();
        fish.swim();
    }

}

File: src/c04_professional_classes/matlab/CapabilityInterfaceDemo.m
function CapabilityInterfaceDemo()
    bird = Bird();
    fish = Fish();

    bird.fly();
    fish.swim();
end

classdef IFlyable < handle
    methods (Abstract)
        fly(obj)
    end
end

classdef ISwimmable < handle
    methods (Abstract)
        swim(obj)
    end
end

classdef Bird < IFlyable
    methods
        function fly(obj)
            fprintf('Bird is flying...\n');
        end
    end
end

classdef Fish < ISwimmable
    methods
        function swim(obj)
            fprintf('Fish is swimming...\n');
        end
    end
end

classdef Duck < ISwimmable & IFlyable
    methods
        function swim(obj)
            fprintf('Duck is swimming...\n');
        end
        function fly(obj)
            fprintf('Duck is flying...\n');
        end
    end
end
